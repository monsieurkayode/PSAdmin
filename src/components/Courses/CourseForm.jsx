import React from 'react';
import {
  string, shape, bool, arrayOf, func
} from 'prop-types';
import TextInput from '../common/TextInput/Index';
import SelectInput from '../common/SelectInput/Index';

/**
 * @function CourseForm
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const CourseForm = ({
  title,
  authorId,
  category,
  length,
  errors,
  saving,
  loading,
  authors,
  onChange,
  onSubmit,
  onFocus
}) => (
  <form onSubmit={onSubmit}>
    <TextInput
      name="title"
      label="Title"
      placeholder="React Fundamentals..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={title}
      error={errors.title}
    />
    <SelectInput
      name="authorId"
      label="Author"
      defaultOption="Select Author"
      onChange={onChange}
      onFocus={onFocus}
      value={authorId}
      loading={loading}
      error={errors.authorId}
      options={authors}
    />
    <TextInput
      name="category"
      label="Category"
      placeholder="UI/UX, HTML, Graphql..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={category}
      error={errors.category}
    />
    <TextInput
      name="length"
      label="Length"
      placeholder="5:12"
      onChange={onChange}
      onFocus={onFocus}
      loading={loading}
      value={length}
      error={errors.length}
    />
    <input
      type="submit"
      value={saving ? 'Saving...' : 'Save'}
      disabled={saving}
      className="btn btn-primary"
    />
  </form>
);

CourseForm.defaultProps = {
  title: '',
  authorId: '',
  category: '',
  length: '',
  errors: {},
  authors: []
};

CourseForm.propTypes = {
  title: string,
  authorId: string,
  category: string,
  length: string,
  errors: shape({}),
  authors: arrayOf(shape({})),
  saving: bool.isRequired,
  loading: bool.isRequired,
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onFocus: func.isRequired
};

export default CourseForm;
