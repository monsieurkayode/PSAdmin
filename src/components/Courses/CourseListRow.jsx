import React from 'react';
import { Link } from 'react-router-dom';
import { string, func } from 'prop-types';

/**
 *
 * @function CourseListRow
 *
 * @param {object} { id, watchHref, title, authorId, category, length }
 *
 * @returns {JSX}
 */
const CourseListRow = ({
  id, watchHref, title, authorId, category, length, deleteCourse
}) => (
  <tr>
    <td>
      <a href={watchHref} target="_blank" rel="noopener noreferrer">
        Watch
      </a>
    </td>
    <td><Link to={`/course/${id}`}>{title}</Link></td>
    <td>{authorId}</td>
    <td>{category}</td>
    <td>{length}</td>
    <td>
      <button
        type="button"
        className="btn btn-sm btn-danger"
        onClick={() => deleteCourse(id)}
      >
        Delete
      </button>
    </td>
  </tr>
);

CourseListRow.propTypes = {
  authorId: string.isRequired,
  category: string.isRequired,
  id: string.isRequired,
  length: string.isRequired,
  title: string.isRequired,
  watchHref: string.isRequired,
  deleteCourse: func.isRequired
};

export default CourseListRow;
