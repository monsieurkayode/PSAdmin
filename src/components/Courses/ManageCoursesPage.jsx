import React, { Component } from 'react';
import {
  shape, bool, arrayOf
} from 'prop-types';
import { connect } from 'react-redux';
import toastr from 'toastr';
import { Prompt } from 'react-router-dom';
import { saveCourse, setSaveErrors } from '../../actions/courseActions';
import CourseForm from './CourseForm';
import NotFoundPage from '../NotFound/NotFoundPage';
import { validateInput, checkChanges } from '../../helpers/validator';

import '../../styles/ManageCoursePage.scss';

/**
 * @class ManageCoursesPage
 *
 * @extends Component
 *
 * @returns {JSX}
 */
class ManageCoursesPage extends Component {
  static defaultProps = {
    course: {}
  }

  static propTypes = {
    isSaving: bool.isRequired,
    course: shape({}),
    authors: arrayOf(shape({})).isRequired,
    loadingCourse: bool.isRequired
  }

  static initialState = () => ({
    course: {
      id: '',
      title: '',
      length: '',
      authorId: '',
      category: '',
      watchHref: '',
    },
    errors: {},
    newChanges: false,
  });

  static getDerivedStateFromProps(
    { course, isSaving }, { errors, newChanges }
  ) {
    if (!isSaving && !newChanges) {
      return {
        ...ManageCoursesPage.initialState(),
        course,
        errors,
        newChanges
      };
    }

    return null;
  }

  state = ManageCoursesPage.initialState();

  redirect = this.props.history.push;

  componentWillUnmount() {
    this.props.setSaveErrors({});
  }

  /**
   * @method clearErrors
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  clearErrors = () => {
    toastr.clear();
    if (Object.keys(this.state.errors).length > 0) {
      this.setState({ errors: {} });
    }
  }

  /**
   * @method handleSubmit
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  handleSubmit = (event) => {
    event.preventDefault();
    const { course } = this.state;
    const { isValid, errors } = validateInput(course);

    this.setState({ newChanges: false });

    if (!isValid) {
      toastr.clear();
      toastr.error('Invalid form data', 'Submission Error!');
      return this.setState({ errors });
    }

    this.props.saveCourse(course)
      .then(() => this.redirect('/courses'))
      .catch(err => this.setState({ errors: err }));
  }

  /**
   * @method handleInputChange
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { course } = this.state;
    course[event.target.name] = event.target.value;
    this.setState({ course, newChanges: checkChanges(course) });
  }

  render() {
    const {
      authors, isSaving, loadingCourse, match
    } = this.props;
    const { course, errors, newChanges } = this.state;

    if (!loadingCourse && match.params.id && !this.props.course.id) {
      return <NotFoundPage content="Course" />;
    }

    return (
      <div className="container" style={{ maxWidth: '55%', minWidth: '400px' }}>
        <Prompt
          when={!isSaving && newChanges}
          message="You have unsaved changes, leave page?"
        />
        <h1>Manage Course</h1>
        <CourseForm
          {...course}
          errors={errors}
          authors={authors}
          onChange={this.handleInputChange}
          onSubmit={this.handleSubmit}
          onFocus={this.clearErrors}
          saving={isSaving}
          loading={loadingCourse}
        />
      </div>
    );
  }
}

/**
 * @function getCourse
 *
 * @param {array} courses
 * @param {string} id
 *
 * @returns {any}
 */
const getCourse = (courses, id) => courses.find(c => c.id === id);

/**
 *
 * @description makes the ManageCoursePage a container component
 * by making it aware of the application state in the store
 * via the connect function from react-redux library
 *
 * @function mapStateToProps
 *
 * @param {object} { allCourses, allAuthors }
 */
const mapStateToProps = ({
  allAuthors: { authors },
  allCourses: {
    courses, isSaving, isLoading, errors
  }
}, { match: { params } }) => {
  let { course } = ManageCoursesPage.initialState();

  if (getCourse(courses, params.id)) {
    course = Object.assign({}, getCourse(courses, params.id));
  }

  const authorsFormattedForDropdown = authors.map(author => (
    {
      value: author.id,
      text: `${author.firstName} ${author.lastName}`
    }
  ));

  let loadingCourse = isLoading;

  if (!params.id) { loadingCourse = false; }

  return {
    isSaving,
    loadingCourse,
    errors,
    courses,
    course,
    authors: authorsFormattedForDropdown
  };
};

export default connect(
  mapStateToProps,
  { saveCourse, setSaveErrors }
)(ManageCoursesPage);
