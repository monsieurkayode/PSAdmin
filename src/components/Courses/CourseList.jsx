import React from 'react';
import { arrayOf, shape, func } from 'prop-types';
import CourseListRow from './CourseListRow';

/**
 *
 * @function CourseList
 *
 * @param {object} props
 *
 * @returns {JSX} JSX
 */
const CourseList = ({ courses, deleteCourse }) => (
  <table className="table">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Title</th>
        <th>Author</th>
        <th>Category</th>
        <th>Length</th>
        <th>&nbsp;</th>
      </tr>
    </thead>
    <tbody>
      {courses.map(course => (
        <CourseListRow
          key={course.id}
          {...course}
          deleteCourse={deleteCourse}
        />
      ))}
    </tbody>
  </table>
);

CourseList.propTypes = {
  courses: arrayOf(shape({})).isRequired,
  deleteCourse: func.isRequired
};

export default CourseList;
