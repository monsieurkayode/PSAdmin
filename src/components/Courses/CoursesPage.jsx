import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
  arrayOf, shape, bool, func
} from 'prop-types';
import { deleteCourse, pageChange } from '../../actions/courseActions';
import CourseList from './CourseList';
import Loader from '../common/Loader/Index';
import Pagination from '../common/Pagination/Index';
import paginator from '../../helpers/paginator';

/**
 *
 * @class CoursesPage
 * @extends Component
 *
 * @returns {JSX}
 */
class CoursesPage extends Component {
  componentWillUpdate({ meta, currentPage }) {
    if (meta.pageSize === 0 && currentPage > 1 && meta.pageCount >= 1) {
      this.props.pageChangeAction(currentPage - 1);
    }
  }

  /**
   * @method handlePageClick
   * @memberof CoursePage
   *
   * @param {object} { selected }
   *
   * @returns {void}
   */
  handlePageClick = ({ selected }) => {
    this.props.pageChangeAction(selected + 1);
  }

  render() {
    const {
      courses,
      isLoading,
      deleteCourseAction,
      meta
    } = this.props;
    return (
      <main>
        <div className="container jumbotron">
          <div className="container-fluid">
            <h1>Courses</h1>
            <p>Courses management page.</p>
            <Link to="/course">
              <button type="button" className="btn btn-lg btn-primary">
                Add Course
              </button>
            </Link>
          </div>
        </div>
        <section className="container">
          <div>
            {isLoading && <Loader size={40} />}
            {!isLoading && courses.length > 0
              && (
              <CourseList
                courses={courses}
                deleteCourse={deleteCourseAction}
              />
              )
            }
            {!isLoading && !courses.length
              && <div className="no-content">The course catalog is empty.</div>
            }
            {!isLoading && courses.length > 0 && (
            <Pagination
              {...meta}
              handlePageChange={this.handlePageClick}
            />
            )}
          </div>
        </section>
      </main>
    );
  }
}

CoursesPage.propTypes = {
  deleteCourseAction: func.isRequired,
  courses: arrayOf(shape({})).isRequired,
  isLoading: bool.isRequired,
  meta: shape({}).isRequired,
  pageChangeAction: func.isRequired
};

/**
 *
 * @description makes the CoursePage a container component
 * by making it aware of the application state in the store
 * via the connect function from react-redux library
 *
 * @function mapStateToProps
 *
 * @param {object} { allCourses }
 */
const mapStateToProps = ({
  allCourses: { courses, isLoading, currentPage }
}) => {
  const sortedCourses = courses.sort(
    (prev, next) => prev.title.toUpperCase() > next.title.toUpperCase()
  );
  const meta = paginator(sortedCourses, currentPage);

  return {
    courses: meta.results,
    isLoading,
    meta,
    currentPage
  };
};

export default connect(
  mapStateToProps,
  { deleteCourseAction: deleteCourse, pageChangeAction: pageChange }
)(CoursesPage);
