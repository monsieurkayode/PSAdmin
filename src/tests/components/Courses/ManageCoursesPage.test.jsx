import React from 'react';
import { shallow } from 'enzyme';
import { ManageCoursesPage } from '../../../components/Courses/ManageCoursesPage';
import mockAuthors from '../../__mocks__/mockAuthors';
import mockCourse from '../../__mocks__/mockCourse';

const setup = () => {
  const props = {
    authors: mockAuthors,
    isSaving: false,
    loadingCourse: false,
    history: {
      push: jest.fn()
    },
    saveCourse: () => new Promise(resolve => resolve(mockCourse)),
    setSaveErrors: jest.fn(),
    course: mockCourse,
    match: {
      params: {
        id: mockCourse.id
      }
    }
  };

  const state = {
    errors: {},
    newChanges: false
  };

  const wrapper = shallow(<ManageCoursesPage {...props} />);

  return {
    wrapper,
    props,
    state
  };
};

const event = {
  target: {
    name: 'title',
    value: ''
  },
  preventDefault: jest.fn()
};

describe('ManageCoursesPage Component', () => {
  const { wrapper, state, props: { isSaving, course } } = setup();

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a h1 tag element', () => {
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toEqual('Manage Course');
  });

  it('renders a Prompt child component', () => {
    expect(wrapper.find('Prompt').length).toBe(1);
  });

  it('renders a CourseForm child component', () => {
    expect(wrapper.find('CourseForm').length).toBe(1);
  });

  it('invokes the static method getDerivedStateFromProps', () => {
    const { errors, newChanges } = state;
    const result = ManageCoursesPage
      .getDerivedStateFromProps({ course, isSaving }, { errors, newChanges });
    expect(result.course).toEqual(course);
  });

  it('invokes handleSubmit method', () => {
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
  });

  it('invokes clearErrors method', () => {
    const clearErrorsSpy = jest.spyOn(wrapper.instance(), 'clearErrors');
    wrapper.instance().clearErrors(event);
    expect(clearErrorsSpy).toHaveBeenCalled();
  });

  it('invokes handleInputChange method', () => {
    const handleInputChangeSpy = jest
      .spyOn(wrapper.instance(), 'handleInputChange');
    wrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();
  });

  it('invokes componentWillMount method', () => {
    jest.spyOn(wrapper.instance(), 'componentWillUnmount');
    wrapper.instance().componentWillUnmount();
    expect(wrapper.instance().componentWillUnmount).toHaveBeenCalled();
  });
});
