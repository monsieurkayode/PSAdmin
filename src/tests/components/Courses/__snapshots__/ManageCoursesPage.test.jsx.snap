// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageCoursesPage Component renders correctly 1`] = `
<div
  className="container"
  style={
    Object {
      "maxWidth": "55%",
      "minWidth": "400px",
    }
  }
>
  <Prompt
    message="You have unsaved changes, leave page?"
    when={false}
  />
  <h1>
    Manage Course
  </h1>
  <CourseForm
    authorId="cory-house"
    authors={
      Array [
        Object {
          "firstName": "Cory",
          "id": "cory-house",
          "lastName": "House",
        },
        Object {
          "firstName": "Scott",
          "id": "scott-allen",
          "lastName": "Allen",
        },
      ]
    }
    category="JavaScript"
    errors={Object {}}
    id="react-flux-building-applications"
    length="5:08"
    loading={false}
    onChange={[Function]}
    onFocus={[Function]}
    onSubmit={[Function]}
    saving={false}
    title="Building Applications in React and Flux"
    watchHref="react-flux-building-applications"
  />
</div>
`;
